### ISSUE TITLE
[Bug] Prefill FLOPs estimate ignores `prefix_lens`, so `est. prefill TFLOPS/s` degenerates into 1/latency across chunked-prefill chunks

### ISSUE THREAD

**Issue by memory-of-star** [sgl-project/sglang#34298]
## Checklist

- [x] I have searched related issues but cannot get the expected help.
- [x] The bug has not been fixed in the latest version (checked `main` @ `77c90e7e54`).
- [x] Please use English, otherwise it will be closed.

## Describe the bug

`SchedulerMetricsReporter._estimate_prefill_perf()` ignores the per-request prefix lengths and collapses all extend tokens in the batch into one fictitious sequence. It only reads `extend_lens` and never consults `batch.prefix_lens`:

https://github.com/sgl-project/sglang/blob/77c90e7e54/python/sglang/srt/managers/scheduler_components/metrics_reporter.py#L456-L480

```python
tokens = max(0, int(sum(batch.extend_lens)))
if tokens == 0:
    return 0.0, 0.0, 0.0

# Causal prefill token-context product.
context_product = tokens * (tokens + 1) / 2.0
flops = (
    tokens * self._linear_flops_per_token
    + self._attn_dot_flops_coeff * context_product
)

read_bytes = (
    tokens * self._weight_read_bytes_per_token
    + tokens * self._qkv_act_bytes_per_token
    + tokens * self._prefill_attn_act_read_per_token
)
```

`extend_lens` is the length of the **current chunk only** — not the cumulative sequence length. See `prepare_for_extend`, where `prefix_lens` and `extend_lens` are built as two separate lists and the invariant is asserted explicitly:

https://github.com/sgl-project/sglang/blob/77c90e7e54/python/sglang/srt/managers/schedule_batch.py#L2364-L2365

```python
prefix_lens = [len(r.prefix_indices) for r in reqs]
extend_lens = [r.extend_range.length for r in reqs]
...
assert seq_len - pre_len == req.extend_range.length   # line 2414
```

So `batch.prefix_lens` is available on the same object and is simply not used by the estimator. There is a second error in the same formula: for a multi-request batch, applying one triangular number to `sum(extend_lens)` invents attention pairs between different requests. For example, two independent 100-token prefills have `2 × 100 × 101 / 2 = 10,100` causal pairs, while the current code estimates `200 × 201 / 2 = 20,100`.

### Consequence 1: with a fixed chunk size the numerator is a constant

The reported value is `flops / gap_latency`:

https://github.com/sgl-project/sglang/blob/77c90e7e54/python/sglang/srt/managers/scheduler_components/metrics_reporter.py#L592-L604

```python
flops, _, _ = self._estimate_prefill_perf(batch)
tflops_per_s = flops / gap_latency / 1e12
msg += f", est. prefill TFLOPS/s (per GPU): {tflops_per_s:.2f}"
```

Because `flops` depends only on `sum(extend_lens)`, every chunk of a long request under a fixed `--chunked-prefill-size` produces the **exact same numerator**. The log line therefore reduces to

```
est. prefill TFLOPS/s (chunk k)  =  C / gap_latency_k        (C constant)
```

i.e. an inverse-latency plot wearing a TFLOPS label. Later chunks generally take longer as their prefix grows, so the number tends to decay even though the GPU is doing more useful work per chunk. Scheduler jitter, concurrency, and the use of an inter-log wall-clock interval mean strict monotonicity is not guaranteed; the deterministic problem is that the numerator remains constant and the metric is proportional only to `1 / gap_latency`. This is the opposite of the intended signal, and it breaks precisely the "trend analysis" use case the docs recommend the metric for:

> These metrics are estimates intended for observability and trend analysis.
> — `docs/docs/references/production_metrics.mdx`

### Consequence 2: the dense-attention undercount grows linearly with prefix length

For a chunk of `c` tokens at prefix `P`, the true causal pair count is `c·P + c(c+1)/2`, while the code computes `c(c+1)/2`. The undercount factor is

```
1 + 2P/(c+1)
```

With `--chunked-prefill-size 8192`, the attention term is undercounted by ~33x at `P` = 128K, ~65x at 256K, and ~129x at 512K. These factors apply to the attention component, not necessarily to the estimator's total FLOPs, which also includes a prefix-independent linear term. The same defect applies to any prefill served from a prefix-cache hit, not just chunked prefill — a request with a 100K-token cache hit and a 200-token extend is scored as a 200-token prefill from scratch.

### Consequence 3: `read_bytes` omits the prefix KV cache read entirely

Every term in the prefill `read_bytes` is `tokens * <per-token constant>`. There is no prefix-dependent KV-cache read term, so prefill contributes no KV-cache read traffic for the cached prefix at all. `sglang:estimated_read_bytes_per_gpu_total` and any MBU derived from it are therefore underestimated for chunked prefill and prefix-cache hits. The size of the error depends on the kernel's tiling and cache reuse, KV layout, cache dtype, and whether the intended metric represents logical traffic or actual HBM traffic.

### Consequence 4: the decode path already does this correctly

`_estimate_decode_perf()` right below it uses the full context for both FLOPs and KV read bytes, which suggests the prefill path is an oversight rather than a deliberate simplification:

https://github.com/sgl-project/sglang/blob/77c90e7e54/python/sglang/srt/managers/scheduler_components/metrics_reporter.py#L482-L506

```python
total_context = float(_decode_total_seq_lens(batch))
flops = (
    tokens * self._linear_flops_per_token
    + self._attn_dot_flops_coeff * total_context
)
read_bytes = (
    ...
    + total_context * self._kv_cache_bytes_per_token
)
```

Note also that this fallback branch is what essentially every deployment hits: `_prefill_sol_suffix()` is an unconditional `return ""` in the base class (line 507) and no subclass overrides it anywhere in the tree, so the `sol_suffix` short-circuit never fires.

## Observed impact

Setup: GLM-5.2 (`GlmMoeDsaForCausalLM`), 8×B300 SXM, TP8, FP8 KV cache, `index_topk=2048`, `--chunked-prefill-size 8192`, `--enable-mfu-metrics`.

Observed behaviour in this run: `est. prefill TFLOPS/s (per GPU)` fell monotonically across the chunks of a single long request, which initially read as "prefill is becoming memory bound at long context". That interpretation is invalid: with a fixed chunk size the estimated numerator is frozen, so changes in the reported value reflect only changes in the inter-log interval. Other runs need not be strictly monotonic because that interval also contains scheduler and concurrency noise.

Analytical/model-based decomposition of one 8192-token chunk (not measured per-kernel timings; absolute times depend on the model configuration and efficiency assumptions):

| prefix `P` | chunk time | MoE+linear | sparse MLA attn | DSA indexer GEMM | logits + top-k | KV dequant | reported TFLOPS (rel.) | achieved TFLOPS (rel.) |
|---:|---:|---:|---:|---:|---:|---:|---:|---:|
| 0 | 56 ms | 15.6 ms | 36.6 ms | 2.3 ms | 0.9 ms | 0.0 ms | 1.00 | 1.00 |
| 32K | 88 ms | 15.6 ms | 41.9 ms | 20.8 ms | 8.5 ms | 1.0 ms | 0.63 | 0.92 |
| 128K | 168 ms | 15.6 ms | 41.9 ms | 76.2 ms | 31.0 ms | 3.3 ms | 0.33 | 0.91 |
| 512K | 489 ms | 15.6 ms | 41.9 ms | 297.9 ms | 121.2 ms | 12.5 ms | 0.11 | 0.90 |

In this model, the reported metric drops ~9x while the modeled achieved FLOP/s remains within 10%. The modeled growth is dominated by the DSA indexer, which the estimator does not account for. This table illustrates the mechanism; the source-level reproduction above does not depend on these modeled timings.

## Additional gaps in `_init_estimated_perf_constants` (same metric, separate root causes)

These are independent of the `prefix_lens` bug but make the absolute magnitude unusable for MLA / MoE / sparse-attention models:

https://github.com/sgl-project/sglang/blob/77c90e7e54/python/sglang/srt/managers/scheduler_components/metrics_reporter.py#L372-L455

1. **MLA is charged using dense MHA/GQA shapes.** `attn_linear_flops` uses `2·hidden·head_dim·(n_q + 2·n_kv) + 2·hidden·head_dim·n_q`. For GLM-5.2 / DeepSeek-V3-class MLA configs, `ModelConfig._derive_model_shapes` sets `head_dim = 256`, while `num_key_value_heads` defaults to `num_attention_heads` when the config does not provide it. This does not represent the actual MLA projection path through fields such as `kv_lora_rank`, `q_lora_rank`, and `qk_rope_head_dim`. For the configuration above it substantially misestimates the projection cost; the magnitude and even the net direction of the full linear-layer error are architecture- and implementation-dependent.

2. **MoE is charged as a dense FFN.** `mlp_flops = 6 · hidden · intermediate_size` uses `hf_text_config.intermediate_size` (the dense width) for all `num_hidden_layers`. The per-token MoE cost is `moe_intermediate_size × num_experts_per_tok` plus shared experts, and `first_k_dense_replace` layers differ from MoE layers. Neither is modeled, so the constant is off by a large factor in an unpredictable direction.

3. **Sparse attention is not modeled.** For DSA/NSA models the main attention is capped at `index_topk` keys per query, so `context_product` should saturate rather than grow — see `DeepseekSparseAttnBackend.forward_extend`, which builds a `page_table` of `index_topk` entries per query token, and `SGLANG_DSA_PREFILL_DENSE_ATTN_KV_LEN_THRESHOLD` (default 2048) which gates the dense path. Conversely, the indexer's `fp8_mqa_logits` scan over the *whole* context is unaccounted for, and because the indexer submodules are `ReplicatedLinear`, that cost is not divided by TP size. At long context this term dominates per-chunk time.

## Reproduction

```bash
python3 -m sglang.launch_server \
  --model-path <any model> \
  --tp-size 8 \
  --chunked-prefill-size 8192 \
  --enable-metrics --enable-mfu-metrics \
  --context-length 262144
```

Send one long request (e.g. 128K input tokens) and watch the `Prefill batch` log lines for that request:

```
Prefill batch, #new-seq: 1, #new-token: 8192, ..., est. prefill TFLOPS/s (per GPU): <A>
Prefill batch, #new-seq: 0, #new-token: 8192, ..., est. prefill TFLOPS/s (per GPU): <B>
Prefill batch, #new-seq: 0, #new-token: 8192, ..., est. prefill TFLOPS/s (per GPU): <C>
...
```

In an otherwise idle single-request run, the values will typically trend downward while `#new-token` stays at 8192 on every line. Strict `A > B > C > ...` ordering is not required to reproduce the estimator bug because `gap_latency` is an asynchronous scheduler interval.

Two checks that isolate the bug from any real hardware effect:

1. Multiplying each line's reported TFLOPS by that line's inter-log interval yields the same product, up to log rounding — confirming the numerator is constant across equal-sized chunks.
2. Send a request with a ~100% prefix-cache hit plus a short extend. The reported TFLOPS is computed purely from the extend length and ignores the cached prefix the attention kernels actually read.

## Suggested fix

Minimal correctness fix — use the `prefix_lens` that is already on the batch, and add the missing KV read term:

```python
def _estimate_prefill_perf(self, batch) -> Tuple[float, float, float]:
    if batch is None or batch.extend_lens is None:
        return 0.0, 0.0, 0.0

    extend_lens = batch.extend_lens
    prefix_lens = batch.prefix_lens or [0] * len(extend_lens)

    tokens = 0
    context_product = 0.0
    context_kv_tokens = 0.0
    for c, p in zip(extend_lens, prefix_lens):
        if c <= 0:
            continue
        tokens += c
        # Causal pairs for this request: chunk-vs-prefix plus chunk-vs-chunk.
        context_product += c * p + c * (c + 1) / 2.0
        # One possible logical-traffic convention: count one stream of the KV
        # context per chunk. This is not a measurement of actual HBM traffic.
        context_kv_tokens += p + c / 2.0

    if tokens == 0:
        return 0.0, 0.0, 0.0

    flops = (
        tokens * self._linear_flops_per_token
        + self._attn_dot_flops_coeff * context_product
    )
    read_bytes = (
        tokens * self._weight_read_bytes_per_token
        + tokens * self._qkv_act_bytes_per_token
        + tokens * self._prefill_attn_act_read_per_token
        + context_kv_tokens * self._kv_cache_bytes_per_token
    )
    ...
```

`prefix_lens` is safe to zip against `extend_lens`: both are assigned together in `prepare_for_extend` and both are appended in parallel by `mix_with_running`, which pushes the running request's full context into `prefix_lens` alongside its `[1] * running_bs` extend entry. So the same patch also fixes mixed prefill+decode batches, where today the appended decode tokens contribute `1` to `context_product` and nothing to the KV read.

The per-request attention pair-count above is exact for standard dense causal self-attention; it is not an exact whole-model FLOPs estimate and does not apply unchanged to sparse attention. The KV `read_bytes` term is only one possible logical-traffic convention, and maintainers should first define what the metric is intended to represent. `p + c/2` approximates one stream of the logical KV context per chunk, while using `context_product` models one KV-vector read per query/key pair, similar to the convention used by `_estimate_decode_perf`. Neither necessarily equals actual HBM traffic: real traffic depends on query tiling, on-chip and L2 reuse, fused paths, the model-specific KV layout, and cache dtype. Either convention at least accounts for the currently omitted prefix, but it should be documented explicitly.

Two follow-ups worth considering, either here or as separate work:

- Cap the per-query context at `index_topk` for sparse-attention models and add the indexer scan term, so DSA/NSA numbers stop being dominated by unmodeled work. `ModelConfig` already exposes `index_head_dim` / the DSA fields needed for this.
- Make `_linear_flops_per_token` MoE- and MLA-aware (`moe_intermediate_size`, `num_experts_per_tok`, `n_shared_experts`, `first_k_dense_replace`, `kv_lora_rank`, `q_lora_rank`), or otherwise document that the absolute value is not meaningful for these architectures.

Separately, dividing by `gap_latency` is documented in-code as a known mismatch (the comment at lines 592-597 notes it is "decoupled from this forward's execution"). Since `_prefill_sol_suffix` never returns content in-tree, that caveat applies to every deployment. Using the device-timer span when `ENABLE_METRICS_DEVICE_TIMER` is set would make the wall-clock branch meaningful too.

## Environment

- SGLang: `main` @ `77c90e7e54` (also present in v0.5.17, v0.5.16, v0.5.15; introduced in #19395 "MFU metrics in Prometheus" and carried through the `SchedulerMetricsReporter` refactor)
- Model: GLM-5.2 NVFP4 (`GlmMoeDsaForCausalLM`), 78 layers, 64 heads, `kv_lora_rank=512`, `qk_rope_head_dim=64`, `index_topk=2048`
- Hardware: 8× B300 SXM, TP8
- The defect is architecture-independent and reviewable from the source alone; the GLM-5.2/DSA details only affect how large the discrepancy gets.

**Comment by Lenoplus42** [comment]
Confirmed, context_product is derived from sum(extend_lens) alone, so a chunk of c tokens sitting at prefix P is scored as a fresh length-c sequence — it drops the c·P query–key pairs against the cached prefix, and collapsing the batch before applying a single triangular number additionally invents pairs between independent requests.

Two things I ran into while tracing the fields, in case they're useful:

- Mixed prefill+decode batches are affected too. mix_with_running appends each running request as extend_len = 1 with prefix_len set to its current context length. Per-request, c·P + c(c+1)/2 yields P + 1 for those rows, but the current code only adds that 1 to the batch-wide token count. The request's actual context length never enters the estimate.
- extend_lens and prefix_lens stay index-aligned on every path that writes them (prepare_for_extend, mix_with_running, prepare_encoder_info_extend), so a plain zip is enough and no length fallback is needed.

Gonna scope to the FLOPs term, with separate regression cases for the prefix contribution and for the cross-request collapse so that each failure mode has its own guard.

**Comment by Lenoplus42** [comment]
Two follow-ups:

read_bytes has no prefix-dependent term at all, estimated_read_bytes_per_gpu_total and any MBU derived from it are understated for chunked prefill and prefix-cache hits.
I have not included a fix for this in the PR, because there are two conventions that diverge here: counting one logical KV stream per chunk and measuring actual HBM traffic give very different numbers once kernel tiling and cache reuse are involved. The choice changes the meaning of an already-exported metric. If a maintainer states which one this is meant to represent, I'm happy to add it in the same PR.

The MLA / MoE / sparse-attention items are a different kind of problem. I'd leave them to someone with the architecture background. Keeping this issue open for them.
